examples/seq2seq/utils.py

```diff
@@ -19,6 +19,27 @@
 from transformers import BartTokenizer
 
 
+def label_smoothed_nll_loss(lprobs, target, epsilon, ignore_index=-100, reduce=True):
+    """From fairseq"""
+    if target.dim() == lprobs.dim() - 1:
+        target = target.unsqueeze(-1)
+    nll_loss = -lprobs.gather(dim=-1, index=target)
+    smooth_loss = -lprobs.sum(dim=-1, keepdim=True)
+    if ignore_index is not None:
+        pad_mask = target.eq(ignore_index)
+        nll_loss.masked_fill_(pad_mask, 0.0)
+        smooth_loss.masked_fill_(pad_mask, 0.0)
+    else:
+        nll_loss = nll_loss.squeeze(-1)
+        smooth_loss = smooth_loss.squeeze(-1)
+    if reduce:
+        nll_loss = nll_loss.sum()  # mean()? Scared to break other math.
+        smooth_loss = smooth_loss.sum()
+    eps_i = epsilon / lprobs.size(-1)
+    loss = (1.0 - epsilon) * nll_loss + eps_i * smooth_loss
+    return loss / lprobs.shape[0], nll_loss / lprobs.shape[0]
```

Comment:
The division should be done by the actual batch size, which should not take the `ignore_index` into account. Using
```
bs = lprobs.shape[0] if ignore_index is None else lprobs.shape[0] - pad_mask.long().sum()
return loss / bs, nll_loss / bs
```
gives the same result as cross entropy for `nll_loss` (and will allow you to uncomment the assert).